rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -499,3 +504,203 @@ pursue [Run Vector Continuously](#run-vector-continuously) in the future but the
+performance profiles for Vector components. Our existing strategy here is ideal
+and we intend no major changes.
+
+#### Push Into the Critical PR Path
+
+Our existing criterion benchmarks do act as a check on our PRs but are not
+mandatory. The feedback loop today is very long, roughly an hour to two hours
+and would provide significant friction for contributors if the check were
+mandatory presently. We must reduce the total time the criterion check takes and
+then, once done, push the criterion benchmarks into the critical path for PRs.
+
+Completion criteria: Without reducing the scope or reliability of our criterion
+benchmarks we intend to complete the full run in a reasonable amount of time,
+say 10 minutes. This implies giving more hardware to the CI process tasked with
+benchmarking and splitting the load across multiple machines.
+
+#### Expand Coverage
+
+Our existing approach to criterion is in a good state. We will be well-served by
+expanding this coverage. At a minimum every source/sink/transform must have a
+benchmark that establishes its bytes-per-second throughput. Additional
+benchmarks that help us determine whether regressions have been introduce and
+where are, of course, for the common good. The source/sink/transforms should be
+tackled in priority order.
+
+Completion criteria: Once every source/sink/transform is covered with criterion
+benchmarks to the standard where we can declare its bytes-per-second throughput
+and guard against regressions in the same this task will be completed.
+
+#### Discover Tolerances
+
+As we expand our criterion coverage we will be able to declare the
+bytes-per-second throughput of our major components. We will include these
+tolerances in our documentation _and_ use these tolerances to calculate the cost
+to run Vector for certain workloads, given the relevant IO and host details. For
+instance, how much does a `syslog -> json_parser -> elasticsearch` Vector cost
+to run if the user is running in GCP us-central1 and expects 1Gb/second of
+syslog per host?
+
+Completion criteria: When a source/sink/transform is covered by criterion
+benchmarks and has a machine discovered throughput characterization we will
+document this for our end users. This task is completed when all tolerances are
+known, documented and find use in higher level documentation for users.
+
+### Test-Harness
+
+The test-harness is our place for integrated performance testing. The results
+are unreliable, infrequent and, while we gain a non-trivial portion of system
+information from the test subjects we do not necessarily have _actionable_
+information from the subjects. These are the concerns we have to repair.
+
+In roughly this order we will do the following.
+
+#### Focus our Efforts
+
+We will pull a single _performance test_ to focus our efforts on. This test will
+be `file -> json_parser -> http`. We will deterministically generate json lines,
```

Comment:
My only concern with the `file` source is that it does not align with the aggregator roadmap, but it is a _very_ important component for agent use cases, something that a lot of large Vector users depend on. This will need explicit consensus given that it does not align with the near-term roadmap though.

Reply:
Yep. I'm very flexible on what the actual target should be, inflexible that we should just pursue one to start. I picked `file` since it seems relatively easy to build something that emits with a regular cadence into a file (or acts as a file). 

Like I say, open to thoughts here.

Reply:
Both `http` or `file` are roughly similar efforts to get a reliable producer built. Starting with either strikes me as very doable.